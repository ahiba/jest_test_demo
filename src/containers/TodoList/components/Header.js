import React, { Component } from 'react'
import { connect } from 'react-redux'
import { actions } from '../store'
// addUndoItem
class Header extends Component {
  constructor(props){
    super(props)
    this.state = {
      value: ''
    }
    this.handleInputKeyUp = this.handleInputKeyUp.bind(this)
  }
  handleInputChange(e) {
    this.setState({
      value: e.target.value
    })
  } 
  handleInputKeyUp(e) {
    const { value } = this.props
    if(e.keyCode === 13 && value) {
      this.props.addUndoItem(value)
      this.props.handleInputChange('')
    }
  }
  render() {
    const { value, handleInputChange } = this.props
    return (
      <div className="header">
        <div className="header_content">
          TodoList
          <input 
            className="header_input"
            data-test="header-input" 
            value={value} 
            onChange={(e) => handleInputChange(e.target.value)}
            onKeyUp={this.handleInputKeyUp}
            placeholder="todo"
          />
        </div>
      </div>
    )
  }
}

const mapState = (state) => {
  return {
    value: state.todo.inputValue
  }
}
const mapDispatch = dispactch => ({
  handleInputChange(value) {
    dispactch(actions.changeInputValue(value))
  }
})

export default connect(mapState, mapDispatch)(Header)