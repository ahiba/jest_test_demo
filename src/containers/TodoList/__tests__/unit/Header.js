import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
// import App from './App';
import Header from '../../components/Header'
import expectExport from '_expect@24.9.0@expect';

Enzyme.configure({ adapter: new Adapter() });
// import ReactDOM from 'react-dom';

it('header 渲染样式正常', () => {
  const wrapper = shallow(<Header />)
  expect(wrapper).toMatchSnapshot()
  // const inputElem = wrapper.find("[data-test='input']")
  // expect(inputElem.length).toBe(1)
  
});

it('header 组件包含一个input 框', () => {
  const wrapper = shallow(<Header />)
  const inputElem = wrapper.find("[data-test='input']")
  expect(inputElem.length).toBe(1)
  
});

it('header 组件 input框内容 初始化应该为空', () => {
  const wrapper = shallow(<Header />)
  const inputElem = wrapper.find("[data-test='input']")
  expect(inputElem.prop('value')).toEqual('')
  
});

it('header 组件 input框内容 当用户输入时跟随变化', () => {
  const wrapper = shallow(<Header />)
  const inputElem = wrapper.find("[data-test='input']")
  // expect(inputElem.prop('value')).toEqual('')
  const userInput = '今天要学习jest'
  inputElem.simulate('change', {
    target: {
      value: userInput
    }
  })
  expect(wrapper.state('value')).toEqual(userInput)
  const newInputElem = wrapper.find("[data-test='input']")
  expect(newInputElem.prop('value')).toBe(userInput)
});

it('header组件 input 框输入回车时,如果input 无内容， 无操作', () => {
  const fn = jest.fn()
  const wrapper = shallow(<Header addUndoItem={fn}/>)
  const inputElem = wrapper.find("[data-test='input']")
  // const value = ''
  wrapper.setState({
    value: ''
  })
  inputElem.simulate('keyUp', {
    keyCode: 13
  })
  expect(fn).not.toHaveBeenCalled()
})

it('header组件 input 框输入回车时,如果input 有内容， 有操作', () => {
  const fn = jest.fn()
  const wrapper = shallow(<Header addUndoItem={fn}/>)
  const inputElem = wrapper.find("[data-test='input']")
  // const value = ''
  wrapper.setState({
    value: '学习测试'
  })
  inputElem.simulate('keyUp', {
    keyCode: 13
  })
  expect(fn).toHaveBeenCalled()
  expect(fn).toHaveBeenLastCalledWith('学习测试')
})

it('header组件 input 框输入回车时,如果input 有内容， 最后应该清除掉', () => {
  const fn = jest.fn()
  const wrapper = shallow(<Header addUndoItem={fn}/>)
  const inputElem = wrapper.find("[data-test='input']")
  // const value = ''
  wrapper.setState({
    value: '学习测试'
  })
  inputElem.simulate('keyUp', {
    keyCode: 13
  })
  const newInputElem = wrapper.find("[data-test='input']")
  expect(newInputElem.prop('value')).toBe('')
  // expect(fn).toHaveBeenCalled()
  // expect(fn).toHaveBeenLastCalledWith('学习测试')
})