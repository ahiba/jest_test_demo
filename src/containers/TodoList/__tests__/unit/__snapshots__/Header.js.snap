// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header 组件 渲染样式正常 1`] = `
<div
  className="header"
>
  <div
    className="header_content"
  >
    TodoList
    <input
      className="header_input"
      data-test="input"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="todo"
      value=""
    />
  </div>
</div>
`;

exports[`header 渲染样式正常 1`] = `
<div
  className="header"
>
  <div
    className="header_content"
  >
    TodoList
    <input
      className="header_input"
      data-test="input"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="todo"
      value=""
    />
  </div>
</div>
`;
